docs/user_guide/adversarial.rst

```diff
@@ -0,0 +1,603 @@
+..
+    In this file, we provide implementation details and example code for
+    adversarial fairness. NOTE: The examples are located at
+    "test_othermlpackages/adversarial_fairness.py" for testing, so any changes
+    to the examples here should be reflected there.
+
+.. _adversarial:
+
+Adversarial Learning
+--------------------
+
+.. currentmodule:: fairlearn.adversarial
+
+Fairlearn provides an implementation of the adversarial
+learning method presented in :footcite:`zhang2018mitigating`.
+We assume we have data :math:`X, A, Y`, where we want to
+predict :math:`Y` from :math:`X` while being fair with respect to :math:`A`
+based on some fairness measure. We firstly create predictor and adversary neural
+networks with learnable weights :math:`W` and :math:`U` respectively. Without
+considering fairness yet, this typical supervised-learning problem aims to
+minimize the predictor loss :math:`L_P`. Now, to improve fairness, we not
+only want to minimize the predictor loss, but we want to decrease the
+adversary's ability to predict the sensitive features from the predictor's
+predictions. Suppose the adversary has loss term :math:`L_A`, then the paper
+trains the predictor with gradient:
+
+.. math::
+    \nabla_W L_P - \text{proj}_{\nabla_W L_A} \nabla_W L_P - \alpha \nabla_W L_A
+
+If this model converges properly,
+the adversary will attain a loss equal to the entropy, so the adversary
+can not
+predict the sensitive features from the predictions.
+Moreover, this model can be trained for either *demographic parity* or
+*equalized odds*. If only the predictions are fed to the adversary, the
+predictor will learn to satisfy demographic parity. If also the true targets are
+fed to the adversary, the predictor will learn to satisfy equalized odds.
+For details, see :footcite:`zhang2018mitigating`.
+
+Firstly, we will dicuss the models that this implementation accepts in
+:ref:`models`.
+Secondly, in :ref:`data_preprocessing`, we discuss the required
+data preprocessing, as :math:`X` must be an array of
+floats. Then, in :ref:`loss_functions`,
```

Comment:
In the current implementation, the user doesn't choose loss functions, so you could either remove the section, or possibly describe the logic that is used to determine the loss function from the data type (e.g., discrete data will result either in binary or multinomial logistic loss, and continuous data into square loss).